addons/binding/org.openhab.binding.hs110/src/main/java/org/openhab/binding/hs110/discovery/HS110DiscoveryService.java

```diff
@@ -0,0 +1,92 @@
+package org.openhab.binding.hs110.discovery;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.TreeMap;
+
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.hs110.HS110BindingConstants;
+import org.openhab.binding.hs110.internal.HS110;
+import org.openhab.binding.hs110.internal.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HS110DiscoveryService extends AbstractDiscoveryService {
+
+    private Logger log = LoggerFactory.getLogger(HS110DiscoveryService.class);
+
+    ///// Network
+    private byte[] discoverbuffer = Util.encryptBytes(HS110.Command.SYSINFO.value);
+    final private DatagramPacket discoverPacket;
+    private boolean willbeclosed = false;
+    private DatagramSocket datagramSocket;
+    private byte[] buffer = new byte[1024];
+    private DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+
+    private final InetAddress broadcast;
+
+    public HS110DiscoveryService() throws IllegalArgumentException, UnknownHostException {
+
+        super(HS110BindingConstants.SUPPORTED_THING_TYPES_UIDS, 2, false);
+        byte[] addr = { (byte) 0xff, (byte) 0xff, (byte) 0xff, (byte) 0xff };
+        broadcast = InetAddress.getByAddress(addr);
+        discoverPacket = new DatagramPacket(discoverbuffer, discoverbuffer.length, broadcast, 9999);
+
+    }
+
+    @Override
+    protected void startScan() {
+
+        try {
+            datagramSocket = new DatagramSocket(null);
+            datagramSocket.setBroadcast(true);
+            datagramSocket.send(discoverPacket);
+            StringBuilder sb = new StringBuilder();
+            for (byte b : discoverPacket.getData()) {
+                sb.append(b);
+            }
+            log.debug(sb.toString());
+            log.debug("Discovery package sent");
+
+            while (true) {
```

Comment:
Definitively not. But i did not recognize a pattern to achieve:
- Spawn a new thread, which loops to receive udp packages in "startScan()"
- _stop_ this thread when scanning should time out (where do i hook in?)